diff --git a/transport/http2_client.go b/transport/http2_client.go
index 63d2c5e0..258695b2 100644
--- a/transport/http2_client.go
+++ b/transport/http2_client.go
@@ -427,9 +427,6 @@ func (t *http2Client) CloseStream(s *Stream, err error) {
 // accessed any more.
 func (t *http2Client) Close() (err error) {
 	t.mu.Lock()
-	if t.state == reachable {
-		close(t.errorChan)
-	}
 	if t.state == closing {
 		t.mu.Unlock()
 		return errors.New("transport: Close() was already called")
