diff --git a/clientv3/remote_client.go b/clientv3/remote_client.go
index b51116305..b8209b8a5 100644
--- a/clientv3/remote_client.go
+++ b/clientv3/remote_client.go
@@ -80,23 +80,21 @@ func (r *remoteClient) tryUpdate() bool {
 	return true
 }
 
-// acquire gets the client read lock on an established connection or
-// returns an error without holding the lock.
 func (r *remoteClient) acquire(ctx context.Context) error {
 	for {
-		r.mu.Lock()
 		r.client.mu.RLock()
 		closed := r.client.cancel == nil
 		c := r.client.conn
+		r.mu.Lock()
 		match := r.conn == c
 		r.mu.Unlock()
-		if c != nil && match {
-			return nil
-		}
-		r.client.mu.RUnlock()
 		if closed {
 			return rpctypes.ErrConnClosed
 		}
+		if match {
+			return nil
+		}
+		r.client.mu.RUnlock()
 		if err := r.reconnectWait(ctx, nil); err != nil {
 			return err
 		}
