diff --git a/hugolib/page.go b/hugolib/page.go
index 578aed9e..f7149ad7 100644
--- a/hugolib/page.go
+++ b/hugolib/page.go
@@ -290,11 +290,10 @@ func (p *Page) initContent() {
 		defer cancel()
 		c := make(chan error, 1)
 
-		p.contentInitMu.Lock()
-		defer p.contentInitMu.Unlock()
-
 		go func() {
 			var err error
+			p.contentInitMu.Lock()
+			defer p.contentInitMu.Unlock()
 
 			err = p.prepareForRender()
 			if err != nil {
@@ -302,12 +301,6 @@ func (p *Page) initContent() {
 				return
 			}
 
-			select {
-			case <-ctx.Done():
-				return
-			default:
-			}
-
 			if len(p.summary) == 0 {
 				if err = p.setAutoSummary(); err != nil {
 					err = p.errorf(err, "failed to set auto summary")
