diff --git a/mvcc/backend/backend.go b/mvcc/backend/backend.go
index eb39bc367..bdc8a0d0f 100644
--- a/mvcc/backend/backend.go
+++ b/mvcc/backend/backend.go
@@ -304,12 +304,7 @@ func (b *backend) defrag() error {
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	// block concurrent read requests while resetting tx
-	b.readTx.mu.Lock()
-	defer b.readTx.mu.Unlock()
-
-	b.batchTx.unsafeCommit(true)
-
+	b.batchTx.commit(true)
 	b.batchTx.tx = nil
 
 	tmpdb, err := bolt.Open(b.db.Path()+".tmp", 0600, boltOpenOptions)
