diff --git a/galley/pkg/fs/fssource.go b/galley/pkg/fs/fssource.go
index 38e61823f..2cf6eea1a 100644
--- a/galley/pkg/fs/fssource.go
+++ b/galley/pkg/fs/fssource.go
@@ -207,8 +207,11 @@ func (s *fsSource) initialCheck() {
 
 // Stop implements runtime.Source
 func (s *fsSource) Stop() {
+	s.fileResorceKeys = nil
+	s.shas = nil
 	close(s.donec)
 	s.watcher.Close()
+	s.donec = nil
 }
 
 func (s *fsSource) process(eventKind resource.EventKind, key, resourceKind string, r *istioResource) {
