diff --git a/pkg/credentialprovider/aws/aws_credentials.go b/pkg/credentialprovider/aws/aws_credentials.go
index 98e6abecf2f..89869e76eaa 100644
--- a/pkg/credentialprovider/aws/aws_credentials.go
+++ b/pkg/credentialprovider/aws/aws_credentials.go
@@ -20,7 +20,6 @@ import (
 	"encoding/base64"
 	"fmt"
 	"strings"
-	"sync"
 	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
@@ -68,7 +67,6 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
 // lazyEcrProvider is a DockerConfigProvider that creates on demand an
 // ecrProvider for a given region and then proxies requests to it.
 type lazyEcrProvider struct {
-	mu             sync.RWMutex
 	region         string
 	regionURL      string
 	actualProvider *credentialprovider.CachingDockerConfigProvider
@@ -123,9 +121,6 @@ func (p *lazyEcrProvider) Enabled() bool {
 // by the client when attempting to pull an image and it will create the actual
 // provider only when we actually need it the first time.
 func (p *lazyEcrProvider) LazyProvide() *credentialprovider.DockerConfigEntry {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-
 	if p.actualProvider == nil {
 		klog.V(2).Infof("Creating ecrProvider for %s", p.region)
 		p.actualProvider = &credentialprovider.CachingDockerConfigProvider{
