diff --git a/pkg/activator/net/revision_backends_test.go b/pkg/activator/net/revision_backends_test.go
index 691a73f00..c6dc31e32 100644
--- a/pkg/activator/net/revision_backends_test.go
+++ b/pkg/activator/net/revision_backends_test.go
@@ -58,7 +58,6 @@ const (
 func revisionCC1(revID types.NamespacedName, protocol networking.ProtocolType) *v1alpha1.Revision {
 	return revision(revID, protocol, 1)
 }
-
 func revision(revID types.NamespacedName, protocol networking.ProtocolType, cc int64) *v1alpha1.Revision {
 	return &v1alpha1.Revision{
 		ObjectMeta: metav1.ObjectMeta{
@@ -97,22 +96,6 @@ func privateSKSService(revID types.NamespacedName, clusterIP string, ports []cor
 	}
 }
 
-func waitForRevisionBackedMananger(t *testing.T, rbm *revisionBackendsManager) {
-	timeout := time.After(200 * time.Millisecond)
-	for {
-		select {
-		// rbm.updates() gets closed after all revisionWatchers have finished
-		case _, ok := <-rbm.updates():
-			if !ok {
-				return
-			}
-		case <-timeout:
-			t.Error("Timed out waiting for revisionBackendManager to finish")
-			return
-		}
-	}
-}
-
 func TestRevisionWatcher(t *testing.T) {
 	logger := TestLogger(t)
 	for _, tc := range []struct {
@@ -430,7 +413,6 @@ func TestRevisionWatcher(t *testing.T) {
 			cancel()
 
 			wg.Wait()
-			assertChClosed(t, rw.done)
 
 			// Autofill out Rev in expectUpdates
 			for i := range tc.expectUpdates {
@@ -440,6 +422,8 @@ func TestRevisionWatcher(t *testing.T) {
 			if got, want := updates, tc.expectUpdates; !cmp.Equal(got, want, cmpopts.EquateEmpty()) {
 				t.Errorf("revisionDests updates = %v, want: %v, diff (-want, +got):\n %s", got, want, cmp.Diff(want, got))
 			}
+
+			assertChClosed(t, rw.done)
 		})
 	}
 }
@@ -714,9 +698,6 @@ func TestRevisionBackendManagerAddEndpoint(t *testing.T) {
 			if got, want := revDests, tc.expectDests; !cmp.Equal(got, want) {
 				t.Errorf("RevisionDests = %v, want: %v, diff(-want,+got):%s\n", got, want, cmp.Diff(want, got))
 			}
-
-			cancel()
-			waitForRevisionBackedMananger(t, rbm)
 		})
 	}
 }
@@ -977,9 +958,6 @@ func TestRevisionDeleted(t *testing.T) {
 	case <-time.After(time.Millisecond * 200):
 		// Wait to make sure the callbacks are executed.
 	}
-
-	cancel()
-	waitForRevisionBackedMananger(t, rbm)
 }
 
 func TestServiceDoesNotExist(t *testing.T) {
@@ -1031,9 +1009,6 @@ func TestServiceDoesNotExist(t *testing.T) {
 		t.Errorf("Unexpected update, should have had none: %v", x)
 	case <-time.After(200 * time.Millisecond):
 	}
-
-	cancel()
-	waitForRevisionBackedMananger(t, rbm)
 }
 
 func TestServiceMoreThanOne(t *testing.T) {
@@ -1098,7 +1073,4 @@ func TestServiceMoreThanOne(t *testing.T) {
 		t.Errorf("Unexpected update, should have had none: %v", x)
 	case <-time.After(200 * time.Millisecond):
 	}
-
-	cancel()
-	waitForRevisionBackedMananger(t, rbm)
 }
