diff --git a/pkg/reconciler/route/route_test.go b/pkg/reconciler/route/route_test.go
index 5a6817551..fc21b1a1b 100644
--- a/pkg/reconciler/route/route_test.go
+++ b/pkg/reconciler/route/route_test.go
@@ -1053,15 +1053,8 @@ func TestUpdateDomainConfigMap(t *testing.T) {
 
 	for _, tc := range expectations {
 		t.Run(tc.expectedDomainSuffix, func(t *testing.T) {
-			ctx, ifs, ctl, watcher, cf := newTestSetup(t)
-			waitInformers, err := controller.RunInformers(ctx.Done(), ifs...)
-			if err != nil {
-				t.Fatalf("Failed to start informers: %v", err)
-			}
-			defer func() {
-				cf()
-				waitInformers()
-			}()
+			ctx, _, ctl, watcher, cf := newTestSetup(t)
+			defer cf()
 			route := getTestRouteWithTrafficTargets(WithSpecTraffic(v1.TrafficTarget{}))
 			route.Name = uuid.New().String()
 			routeClient := fakeservingclient.Get(ctx).ServingV1().Routes(route.Namespace)
